### PR TITLE
Case Study 1 - extra revisions

### PR DESCRIPTION
**Pull request by cc-a**
As with https://github.com/ImperialCollegeLondon/digital_research_sustainability_training/pull/88 I started reviewing -85 and was finding it hard to put it in a review format so ended up just taking a pass at it myself.

This mostly shuffles around the structure and adds some more concrete pre- and post estimates of emissions.

### CODE REVIEW

File: episodes/04_case_study1.md [hunk 1]
```diff
@@ -163,49 +187,77 @@ optimising](https://step-up.ac.uk/blog/2026-03-31-profile-optimise-py/) her code
...
 
-### Celia Takes Action
+## Celia Takes Action
+
+Celia identifies several ways she can improve the emissions associated with usage
+of her code base.
+
+### Code Optimisation
+
+Optimize the code itself to make more effective use of available computational
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md]:
```suggestion
Optimise the code itself to make more effective use of available computational
```

File: episodes/04_case_study1.md [hunk 2]
```diff
@@ -163,49 +187,77 @@ optimising](https://step-up.ac.uk/blog/2026-03-31-profile-optimise-py/) her code
...
-Finally, she explores the product carbon footprint (PCF) data for her laptop to find the
-embedded emissions from the hardware components. She finds that the total embedded
-emissions from the CPU, GPU, and RAM is 100 kg CO2e.
 :::::::::::::::::::::::::::::::::::::::::::
 
-## References
+## Outcomes
+
+<!-- markdownlint-disable-next-line line-length -->
+![Carbon emissions for different research actions comparing pre- and post-intervention](fig/case_study1_outcomes.png){alt='A bar chart comparing the emissions from Software Development, GitHub Actions, LLM usage and Software usage before and after implementation of emissions reduction measures'}
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md:262]: I am missing where this plot is coming from.

Also the title says "Case Study 3 Outcomes" we might want to change it to Case Study 1?

File: episodes/04_case_study1.md [hunk 3]
```diff
@@ -31,127 +33,149 @@ Celia is concerned about the environmental impact of her software package.
...
-5 workflows that run on GitHub Actions, and they run around 10 times a day.
+### What data does Celia need to understand the emissions associated with her software?
+
+- What are the key aspects of her work that Celia could estimate emissions for?
+- What methodologies could Celia use to estimate her emissions?
+- What additional data would she need to collect in each case?
+
+::::::::::::::::::::::::::::::: solution
+
+1. Use of laptop for development. Embodied emissions of the laptop should be readily
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md]:
```suggestion
1. **Use of laptop for software development**: Embodied emissions of the laptop should be readily
```

File: episodes/04_case_study1.md [hunk 4]
```diff
@@ -31,127 +33,149 @@ Celia is concerned about the environmental impact of her software package.
...
+   findable in a PCF sheet. As her laptop underpins all of her work however it wouldn't
+   be appropriate to ascribe the full embodied emissions to her package. She could
+   ascribe a proportion of the embodied emissions to development of the package.
+   Regarding the operational emissions she could attempt direct measurement of her
+   laptop with a power meter but given the proportionally low operational emissions of
+   consumer electronic devices a rough calculation using the Green Algorithms Calculator
+   would likely be sufficient.
+1. Use of LLMs. Tracking which model she uses, and how many queries she sends and the
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md]:
```suggestion
1. **Use of LLMs**: Tracking which model she uses, and how many queries she sends and the
```

File: episodes/04_case_study1.md [hunk 5]
```diff
@@ -31,127 +33,149 @@ Celia is concerned about the environmental impact of her software package.
...
+   approximate size of the replies for use with the Hugging Face Ecologits calculator.
+1. Execution of the package. Celia's software is used by many different users on a
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md]:
```suggestion
1. **Execution of the package**: Celia's software is used by many different users on a
```

File: episodes/04_case_study1.md [hunk 6]
```diff
@@ -31,127 +33,149 @@ Celia is concerned about the environmental impact of her software package.
...
+   variety of systems. It's unlikely to be practical to get direct energy measurements
+   in this case. More practically she could make use of the [Green Algorithms
+   Calculator]. She should try and get as much detail as possible from users about how
+   much they use the software and with what hardware. She's unlikely to be able to get
+   information from all users but if she can get a representative sample then she could
+   extrapolate to get a rough estimate. Estimating the embodied emissions for all of the
+   machines the code is running on would be a potentially large job so she might want to
+   put that out of scope.
+1. Use of GitHub actions. Information about how often her workflows run and how long
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md]:
```suggestion
1. **Use of GitHub actions**: Information about how often her workflows run and how long
```

File: episodes/04_case_study1.md [hunk 7]
```diff
@@ -31,127 +33,149 @@ Celia is concerned about the environmental impact of her software package.
...
-She finds that they are using it on a local server with 16 CPU cores and 64 GB of RAM.
-They run the package for around 4 hours per week.
+They are able to provide her with the full specification of the machine they are using -
+an [HP Z2 Tower G1i Workstation]. They run the package for around 18 hours a week using
+all 20 cores.
+
+[HP Z2 Tower G1i Workstation]: https://www.hp.com/gb-en/shop/product.aspx?id=8T229EA&opt=ABU&sel=DTP
+
+From individuals she's in contact with, conversations she's had at conferences, mentions
+in academic papers and a workshop she ran recently Celia estimates that her code has
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md]:
```suggestion
in academic papers and a workshop she ran recently, Celia estimates that her code has
```

**Review by SaranjeetKaur — APPROVED** [review]
I think this is in nice shape now. Thanks @cc-a!

Left some minor edits here and there.

In general, I wonder if we should have a keypoints section (in addition to the outcomes section) at the end of each case study to highlight the main takeaways? 
The outcomes section can be worded as a continuation of the case study (in this case, referring to Celia) whereas, the keypoints section can be more generic applicable to any RSE

File: episodes/04_case_study1.md [hunk 8]
```diff
@@ -31,127 +33,149 @@ Celia is concerned about the environmental impact of her software package.
...
-with the development and usage of her software package, what are some measures that
-she can take to reduce these emissions?
-
-- Are there any ways to make the code base more efficient?
-- How can she ensure that the current and future users of her package are using it
-in an efficient way?
-- Any measure she can take to reduce the emissions associated with the use of GitHub Actions?
-- How can she track the embedded emissions from the hardware (laptop) she uses for
-software development?
+What measures can Celia take to redure to reduce the emissions from usage of her
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md]:
```suggestion
What measures can Celia take to reduce the emissions from the usage of her
```

File: episodes/04_case_study1.md [hunk 9]
```diff
@@ -31,127 +33,149 @@ Celia is concerned about the environmental impact of her software package.
...
-- Model used
-- Memory available in GB
-- Platform used for the software development (e.g. local server, personal computer,
-cloud computing)
-- Location to retrieve the energy mix of the location
-- Real usage factor of the CPU
-- Power Usage Efficiency (PUE) of the local data centre (if applicable)
-- Any multiplicative factor to use
+- Software Development - From the model of her laptop she is able to find the PCF
+  datasheet from the manufacterer - [HP EliteBook 840 G9 PCF Sheet]. This gives a total
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md]:
```suggestion
  datasheet from the manufacturer - [HP EliteBook 840 G9 PCF Sheet]. This gives a total
```

File: episodes/04_case_study1.md [hunk 10]
```diff
@@ -163,49 +187,77 @@ optimising](https://step-up.ac.uk/blog/2026-03-31-profile-optimise-py/) her code
...
+
+Celia takes a pass at improving the error handling and input validation in her code to
+reduce the likelihood of running into errors that lead to repeated runs of the code. She
+implements a new configuration validation approach. She ways to catch some failure modes
+early before significant computation has occured.
+
+Again it's difficult to estimate the impact of this work but when these changes were
+released Celia is contacted by several users confused by the new errors. This suggests
+the changes are catching at least some errors.
+
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md:244]:
```suggestion

### Hardware Usage

She decides to keep using the laptop for as long as its lifespan, instead of replacing it too soon. 

```

File: episodes/04_case_study1.md [hunk 11]
```diff
@@ -163,49 +187,77 @@ optimising](https://step-up.ac.uk/blog/2026-03-31-profile-optimise-py/) her code
...
+Celia identifies several ways she can improve the emissions associated with usage
+of her code base.
+
+### Code Optimisation
+
+Optimize the code itself to make more effective use of available computational
+resources. Celia uses a profiler with her code to identify areas where the code could be
+optimised. She identifies the areas of the code where the bulk of the computation is
+performed. After some experimentation she finds a way to improve use of SIMD in a key
+calculation. In her tests this gives a 5% performance boost to the code.
```
Comment (SaranjeetKaur) [inline episodes/04_case_study1.md]:
```suggestion
calculation. Furthermore, she replaces the use of `pandas` with `polars` and reverses the order of a conditional statement and a loop deep into the code, so that the former is not checked several times unnecessarily. In her tests this gives a 7% performance boost to the code.
```

Comment (SaranjeetKaur) [reply episodes/04_case_study1.md]: Suggestion based on Diego's comment: https://github.com/ImperialCollegeLondon/digital_research_sustainability_training/pull/85#discussion_r3128556814

File: episodes/04_case_study1.md (hunk 10 above)
Comment (SaranjeetKaur) [reply episodes/04_case_study1.md:244]: Suggestion based on Diego's comment: https://github.com/ImperialCollegeLondon/digital_research_sustainability_training/pull/85#discussion_r3128542111

File: episodes/04_case_study1.md (hunk 2 above)
Comment (cc-a) [reply episodes/04_case_study1.md:262]: Thanks. Will update and add the script used to generate it.
